lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -0,0 +1,376 @@
+        super().__init__(handle)
+        self.message = message
+
+    def snapshot(self) -> Dict:
+        """Save grafana source information."""
+        return {"message": self.message}
+
+    def restore(self, snapshot):
+        """Restore grafana source information."""
+        self.message = snapshot["message"]
+
+
+class K8sResourcePatchEvents(ObjectEvents):
+    """Events raised by :class:`K8sResourcePatchEvents`."""
+
+    patch_failed = EventSource(K8sResourcePatchFailedEvent)
+
+
+class ContainerNotFoundError(ValueError):
+    """Raised when a given container does not exist in the list of containers."""
+
+
+class KubernetesComputeResourcesPatch(Object):
+    """A utility for patching the Kubernetes compute resources set up by Juju."""
+
+    on = K8sResourcePatchEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        container_name: str,
+        *,
+        limits: Optional[ResourceSpecDict],
+        requests: Optional[ResourceSpecDict],
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
+    ):
+        """Constructor for KubernetesComputeResourcesPatch.
+
+        References:
+            - https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+
+        Args:
+            charm: the charm that is instantiating the library.
+            container_name: the container for which to apply the resource limits.
+            limits: a dictionary for `limits` resources.
+            requests: a dictionary for `requests` resources.
+            refresh_event: an optional bound event or list of bound events which
+                will be observed to re-apply the patch.
+        """
+        super().__init__(charm, "kubernetes-compute-resource-patch")
+        self._charm = charm
+        self._container_name = container_name
+        self.resource_reqs = ResourceRequirements(
+            limits=sanitize_resource_spec_dict(limits),  # type: ignore[arg-type]
+            requests=sanitize_resource_spec_dict(requests),  # type: ignore[arg-type]
+        )
+        self.patched_delta = self._patched_delta(
+            namespace=self._namespace,
+            app_name=self._app,
+            container_name=self._container_name,
+            resource_reqs=self.resource_reqs,
+        )
+
+        # Make mypy type checking happy that self._patch is a method
+        assert isinstance(self._patch, MethodType)
+        # Ensure this patch is applied during the 'config-changed' event, which is emitted every
+        # startup and every upgrade. The config-changed event is a good time to apply this kind of
+        # patch because it is always emitted after storage-attached, leadership and peer-created,
+        # all of which only fire after install. Patching the statefulset prematurely could result
+        # in those events firing without a workload.
+        self.framework.observe(charm.on.config_changed, self._patch)
+
+        if not refresh_event:
```

Comment:
Can the `refresh_event` argument to the constructor be a "tuple" of `BoundEvent` with a default value of an empty tuple ? If so would this simplify this could logic by eliminating the conditional statement ?

Reply:
Our understanding use cases here has been evolving. So if you think using a non mutable vector default argument makes sense, I see no harm in making Prometheus and Grafana consistent with your choice.